autogen/img_utils.py

```diff
@@ -298,3 +310,51 @@ def message_formatter_pil_to_b64(messages: List[Dict]) -> List[Dict]:
         new_messages.append(message)
 
     return new_messages
+
+
+def is_vision_model(model_name: str) -> bool:
+    """
+    Determines if a given model name indicates a multimodal model.
+
+    Multimodal models are identified by specific patterns in their names:
+    models that include "gpt-<variant>-vision" or "llava" are considered
+    multimodal.
+
+    Args:
+        model_name (str): The name of the model to be checked.
+
+    Returns:
+        bool: True if the model is multimodal, False otherwise.
+    """
+    if not isinstance(model_name, str):
+        return False
+    if re.findall(r"gpt-\w+-vision", model_name):
+        return True
+    if re.findall("llava", model_name):
+        return True
+    return False
+
+
+def format_message_contents_with_images(messages: List[Dict]) -> List[Dict]:
+    """
+    Processes a list of message dictionaries, applying formatting to the
+    content of each message if applicable.
+
+    Each message's content is formatted using gpt4v_formatter when the content is a string;
+    so, images in the messages will be converted into a specific format (e.g., URI).
+    If the message is already an array (as the OpenAI message multimodal content), it will
+    be left unchanged so that OpenAI client can handle directly.
+
+
+    Args:
+        messages (List[Dict]): A list of message dictionaries to be processed.
+
+    Returns:
+        List[Dict]: A new list of processed message dictionaries.
+    """
+    rst = []
+    for message in messages:
+        rst.append(copy.deepcopy(message))
```

Comment:
If the message contains an image, we would use the `gpt4v_formatter` function to update the "content" of the message. 

Because the message here is a dict (a pointer to the memory), the underlying `messages` would be modified in the agent.

Reply:
Why isn't `copy()` enough? The modification is in the "content" field only for rst[-1]. It won't alter the original `message` variable.

Reply:
Yes. copy() would also work. Updated